apis/console/src/tests.rs

```diff
@@ -69,3 +99,33 @@ fn failed_print() {
     // The fake driver still receives the command even if a fake error is injected.
     assert_eq!(driver.take_bytes(), b"abcde");
 }
+
+#[test]
+fn failed_read() {
+    let kernel = fake::Kernel::new();
+    let driver = fake::Console::new_with_input(b"bugxxxx");
+    kernel.add_driver(&driver);
+    kernel.add_expected_syscall(ExpectedSyscall::AllowRw {
+        driver_num: DRIVER_NUM,
+        buffer_num: allow_rw::READ,
+        return_error: None,
+    });
+    kernel.add_expected_syscall(ExpectedSyscall::Subscribe {
+        driver_num: DRIVER_NUM,
+        subscribe_num: subscribe::READ,
+        skip_with_error: None,
+    });
+    kernel.add_expected_syscall(ExpectedSyscall::Command {
+        driver_id: DRIVER_NUM,
+        command_id: command::READ,
+        argument0: 3,
+        argument1: 0,
+        override_return: Some(command_return::failure(ErrorCode::Fail)),
+    });
+
+    let mut buf = [0; 3];
+
+    let (count, res) = Console::read(&mut buf);
+    assert_eq!(res, Err(ErrorCode::Fail));
+    assert_eq!(count, 0);
```

Comment:
The PR description mentioned some sort of issue, but it's unclear what that issue is. Are you saying that if you put an `assert_eq!(buf, b"bug");` here it fails? I would expect `buf` to contain `bug` here.

Reply:
I was surprised that count = 0. It doesn't make a lot of sense to have 0 count but 3 chars actually copied, so I made a note that there might be a problem. But since it only possibly affects the test mock, it's out of scope for this change so I did not investigate.

I checked and "bug" is indeed copied.

Reply:
I see. Here's what happens:

* `Console::read` calls the Command system call.
* `fake::Syscalls` calls `<fake::Console as fake::SyscallDriver>::command`, which copies the data and schedules an upcall indicating a copy of 3 bytes was performed. Note that `fake::Console` is unaware of the `ExpectedSyscall`, so it doesn't know it is supposed to return an error.
* `fake::Syscalls` then overrides the return value from `<fake::Console as fake::SyscallDriver>::command` with the error from the `ExpectedSyscall`, returning an error over the system call interface.
* `Syscalls::command` returns an error, which causes the closure passed to `share::scope` in `Console::read` to exit.

`bytes_received` is never changed, so `Console::read` returns `0`.

The `ExpectedSyscall` mechanism isn't flexible enough to simulate "this read failed and didn't copy data" errors. I'd recommend adding a way to inject fake errors into `fake::Console` instead of using the `ExpectedSyscall` mechanism of `fake::Kernel`.